### PR TITLE
Fix context keys get set

### PR DESCRIPTION
**Pull request by rohan-j-shah**
Fix for : fatal error: concurrent map read and map write

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -160,14 +162,18 @@ func (c *Context) Set(key string, value interface{}) {
 	if c.Keys == nil {
 		c.Keys = make(map[string]interface{})
 	}
+	c.KeysLocker.Lock()
```
Comment (cookingkode) [inline context.go]: I would suggest taking it before nil check to avoid a set being lost..

**Comment by maxatome** [comment]
The locking mechanism involved in this pull request is not needed as the original gin.Context should not be used outside the original gin gonic goroutine without copying it (Copy method). In fact, it even hides some biggest errors!
If you have a "concurrent map read and map write" error, it means that you did a mistake. Probably using the same context in different goroutines without copying it.
If you need to share a gin.Context between several goroutines of your own, you first have to copy (Copy method) it then handle the locking mechanism yourself at your level, perhaps by composing the gin.Context in your own struct.
The problem with your patch is that it hides completely the fact that gin.Context instances are reused once the request is handled at the gin gonic level.

**Comment by tsirolnik** [comment]
@maxatome I'm not sharing the context but I still do get panics sometimes even though the key was set.

See this issue #836

**Comment by thinkerou** [comment]
@rohan-j-shah please re-commit the pull request to `master` branch, thanks!

**Comment by thinkerou** [comment]
closing, ref #1391

**Pull request closed without merging**